add filter by group or rule name to vmalert ui

victoramsantos opened the pull request:
### Is your feature request related to a problem? Please describe

At my company we run hundreds of alerts in different groups and to search for them in the vmalert UI sometimes is tough task.  

### Describe the solution you'd like

I'm proposing here a filter at UI level by Group or Rule names and by Labels. The idea is just to show Groups that match its name with the filter and hide the others. For Rule name, I want to filter them and show theirs groups and already show the rules row opened. For labels, I want to filter by its key or its value.

Example:
No filter applied:

<img width="550" alt="Screenshot 2024-02-09 at 14 51 54" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/1914482/2319450a-becb-4952-a6c4-652d88546a44">

Filter by group name `1abc`:

<img width="509" alt="Screenshot 2024-02-09 at 14 52 18" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/1914482/079ac96c-4f6d-43a2-9e61-a3dc23a2297f">

Just changing the query to `2abc`:

<img width="532" alt="Screenshot 2024-02-09 at 14 52 57" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/1914482/4bfb668e-93a9-4262-943b-f1cae75a3c11">

Now, changing to filter by rule name and querying `1234`:

<img width="679" alt="Screenshot 2024-02-09 at 14 53 22" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/1914482/a8b2e520-3859-4018-b514-90daaa68c508">

Changing the query to `5678`:

<img width="697" alt="Screenshot 2024-02-09 at 14 53 48" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/1914482/ed3637b2-c40b-49f2-82b0-3ec0dcfd5298">

Changing the query to `service1` which is a value of a label:

<img width="710" alt="Screenshot 2024-02-09 at 14 54 28" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/1914482/0c5dac04-2583-49de-bac9-dcf772c246ee">

Changing the query to `raw_name=alert_cba`:

<img width="726" alt="Screenshot 2024-02-09 at 14 55 24" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/1914482/f0901823-0802-4655-8776-14ce6412d4c5">

Other case:
- Any blank filter will show all groups
- Every new query show all groups and rules, to ensure we are not hiding wrong elements

Sample alerts used: https://gist.github.com/victoramsantos/1f231ecf879a77e4c4ac181a9113ada2

### Describe alternatives you've considered

I though to run the filter for every letter typed but I think for huge list of groups or rules this could be bad, so I just leave the `Filter` button.

### Additional information

I'm not good with CSS so if someone could help me to beautify the UI would be good 😊 

**Any feedback is welcome 😊**

victoramsantos commented:
Request for review and feedbacks about the feature idea: @Haleygo @hagen1778

hagen1778 commented:
@victoramsantos thanks for pull request! What do you think about having just a text area for search and filter by group/rule names/labels/annotations everything user enters? I checked Prometheus UI, and it does exactly this - see http://demo.robustperception.io:9090/alerts?search=critical
I believe, this is more intuitive for user than radio buttons.

> I though to run the filter for every letter typed but I think for huge list of groups or rules this could be bad, so I just leave the Filter button.

I think this one is easy to fix - just add a delay since last key-up event before executing the search.

I'm cc'ing @Loori-R and @dmitryk-dk here since they have much more experience with JS than I do. Would love to hear what you think guys - can we add responsive search on this page? Should we add something similar to vmagent UI?

victoramsantos commented:
Nice @hagen1778, but I think it's only filtering by rule name, label key and label value. Any search for some string in `description` is returning nothing.
I will add this behavior too.

dmitryk-dk commented:
> @victoramsantos thanks for pull request! What do you think about having just a text area for search and filter by group/rule names/labels/annotations everything user enters? I checked Prometheus UI, and it does exactly this - see http://demo.robustperception.io:9090/alerts?search=critical I believe, this is more intuitive for user than radio buttons.
> 
> > I though to run the filter for every letter typed but I think for huge list of groups or rules this could be bad, so I just leave the Filter button.
> 
> I think this one is easy to fix - just add a delay since last key-up event before executing the search.
> 
> I'm cc'ing @Loori-R and @dmitryk-dk here since they have much more experience with JS than I do. Would love to hear what you think guys - can we add responsive search on this page? Should we add something similar to vmagent UI?

Hi! I have checked the http://demo.robustperception.io:9090/alerts?search=critical and it works nice. I think we can add debounce on every key up event and filter it via search field instead of the checkbox. Because there can be many different filters in future and add another checkbox would make the UI not clear

victoramsantos commented:
Hi, @hagen1778 and @dmitryk-dk I changed the code 😊 
- Now it is just querying by typing. 
- I also added a debounce function (I find one online so I don't know if its the best way to do this :/). 
- It's now querying the group name, rule name, label key and label value.

Loori-R commented on app/vmalert/static/js/custom.js:
```diff
@@ -15,6 +17,93 @@ function toggleByID(id) {
+    };
+};
+
+$('#filter').on("keyup", (debounce(function(){
+    console.log("type")
+    filter();
+},500)));
```
```suggestion
$('#filter').on("keyup", debounce(filter, 500));
```

Loori-R commented on app/vmalert/static/js/custom.js:
```diff
@@ -15,6 +17,93 @@ function toggleByID(id) {
+            let target = $(this).attr("data-bs-target");
+            
+            $(this).show();
+            $("div[id='"+target+"'] .rule").show();
```
```suggestion
            $(`div[id="${target}"] .rule`).show();
```

Loori-R commented on app/vmalert/static/js/custom.js:
```diff
@@ -15,6 +17,93 @@ function toggleByID(id) {
+        }else{
+            let target = $(this).attr('data-bs-target')
+
+            $("div[id='rules-"+target+"'").addClass('show');
+            $("div[data-bs-target='rules-"+target+"'").show();
```
```suggestion
            $(`#rules-${target}`).addClass('show');
            $(`div[data-bs-target='rules-${target}']`).show();
```

Loori-R commented on app/vmalert/static/js/custom.js:
```diff
@@ -15,6 +17,93 @@ function toggleByID(id) {
+        if (matches > 0){
+            let target = $(this).attr('data-bs-target')
+
+            $("div[id='rules-"+target+"'").addClass('show');
+            $("div[data-bs-target='rules-"+target+"'").show();
```
```suggestion
            $(`#rules-${target}`).addClass('show');
            $(`div[data-bs-target='rules-${target}']`).show();
```

Loori-R reviewed (commented):
Overall, it looks good! I've added a few comments on the code style.

dmitryk-dk commented on app/vmalert/static/js/custom.js:
```diff
@@ -15,6 +17,93 @@ function toggleByID(id) {
 }
 
+//http://davidwalsh.name/javascript-debounce-function
+function debounce(func, wait, immediate) {
+    var timeout;
+    return function() {
+        var context = this, args = arguments;
+        var later = function() {
+            timeout = null;
+            if (!immediate) func.apply(context, args);
+        };
+        var callNow = immediate && !timeout;
+        clearTimeout(timeout);
+        timeout = setTimeout(later, wait);
+        if (callNow) func.apply(context, args);
+    };
+};
```
I think we can simplify this debounce function

```javascript
function debounce(func, delay) {
    let timer;
    return function(…args) {
        clearTimeout(timer);
        timer = setTimeout(() => {
            func.apply(this, args);
        }, delay);
    };
}
```

victoramsantos replied:
Nice!

dmitryk-dk reviewed (commented):
Have some comment

victoramsantos commented:
Accepted all comments. 
@hagen1778 for your review 😊

Loori-R reviewed (approved):
LGTM! Thank you!

dmitryk-dk commented on app/vmalert/static/js/custom.js:
```diff
@@ -15,6 +17,84 @@ function toggleByID(id) {
+
+function filterGroupsByName(){
+    $( ".group-heading" ).each(function() {
+        var groupName = $(this).attr('data-group-name');
+        var filter = $("#filter").val()
+
```
One more little thing. you are using `let` in the parts of code, so I think it is better to use const instead of `var`.
Also please check your `let` usage. In some cases, as I can see you do not change the value, so probably it is `const` variable

victoramsantos replied:
Nice tip!

dmitryk-dk commented on app/vmalert/static/js/custom.js:
```diff
@@ -15,6 +17,84 @@ function toggleByID(id) {
+    $(".rule-table").removeClass('show');
+    $(".rule").show();
+    
+    if($("#filter").val().length === 0){
+        $(".group-heading").show()
+        return
+    }
```
```javascript
const hasValue = $("#filter").val().length === 0;
if (hasValue) {
...
}
```

dmitryk-dk commented on app/vmalert/static/js/custom.js:
```diff
@@ -15,6 +17,84 @@ function toggleByID(id) {
+
+function filterRuleByName(){
+    $( ".rule" ).each(function() {
+        var ruleName = $(this).attr("data-rule-name");
+        var filter = $("#filter").val()
```
the same about `const` here

dmitryk-dk commented on app/vmalert/static/js/custom.js:
```diff
@@ -15,6 +17,84 @@ function toggleByID(id) {
+        var ruleName = $(this).attr("data-rule-name");
+        var filter = $("#filter").val()
+        
+        if (ruleName.indexOf(filter) < 0){
```
```javascript
const contains = ruleName.indexOf(filter) > 0;

if (contains) {
 branch about contains
} else {
 branch if not contains
}
```

dmitryk-dk reviewed (commented):
Overall looks good! Added some cleanup comments

dmitryk-dk reviewed (commented):
I have checked how it works and have some other comments.

First of all, could you prettify the content, please? 
The search field should have some margin on top because it is near the buttons. The second moment - it has full width. Could you make it the same width as the buttons?
Could you add the search icon to the left or right side of the search input as it was implemented in the demo? 

Overall, thank you for your work!

<img width="310" alt="Screenshot 2024-02-16 at 09 50 18" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/12189570/db96e511-81a7-41b6-98e5-50fdee69699b">

<img width="1680" alt="Screenshot 2024-02-16 at 09 46 18" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/12189570/b8f56b01-c780-49f3-81cd-78b47a53ffc6">

victoramsantos commented:
Hi, I think now looks good 😊 
WDYT @dmitryk-dk ?
<img width="1723" alt="Screenshot 2024-02-16 at 10 02 44" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/1914482/7133ea23-182a-46c8-a6e4-3260e76681dd">

dmitryk-dk commented:
> Hi, I think now looks good 😊 WDYT @dmitryk-dk ? <img alt="Screenshot 2024-02-16 at 10 02 44" width="1723" src="https://private-user-images.githubusercontent.com/1914482/305421817-7133ea23-182a-46c8-a6e4-3260e76681dd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDgwODk2NTQsIm5iZiI6MTcwODA4OTM1NCwicGF0aCI6Ii8xOTE0NDgyLzMwNTQyMTgxNy03MTMzZWEyMy0xODJhLTQ2YzgtYTZlNC0zMjYwZTc2NjgxZGQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMTZUMTMxNTU0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9N2M0YTczY2I0NDNlMjhkMTkyZGJjYjJkOTkwY2U5Yjk0NDVlOGQ1YmRkNmJjODRhZmQyNTM3OTRkNjdhNjA1MCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.wpFikQiaB3YXGIJOJBww8mXFycb4DNMGj8Kumms-Cik">

One small thing ! Could you move the search field under the button?

victoramsantos commented:
@dmitryk-dk like this?
<img width="862" alt="Screenshot 2024-02-16 at 11 05 51" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/1914482/9383027f-438d-40e1-9a45-1612dabf2751">

dmitryk-dk reviewed (approved):
LGTM!

victoramsantos commented:
@hagen1778 for your review 😊

hagen1778 commented:
Thank you @victoramsantos !

hagen1778 commented:
@victoramsantos please see follow-up https://github.com/VictoriaMetrics/VictoriaMetrics/commit/11b03d9fc8374453cf0a132597598138ae420d34

valyala commented:
FYI, this pull request has been included in [vmalert v1.99.0](https://docs.victoriametrics.com/changelog/).
